profile-widget/diveprofileitem.cpp

```diff
@@ -161,37 +166,89 @@ void DiveProfileItem::replot(const dive *d, int from, int to, bool in_planner)
 	pat.setColorAt(0, getColor(DEPTH_TOP));
 	setBrush(QBrush(pat));
 
-	int last = -1;
-	for (int i = from; i < to; i++) {
-		struct plot_data *pd = dataModel.data().entry;
-		struct plot_data *entry =  pd + i;
-		// "min/max" are the 9-minute window min/max indices
-		struct plot_data *min_entry = pd + entry->min;
-		struct plot_data *max_entry = pd + entry->max;
-
-		if (entry->depth < 2000)
-			continue;
+	// No point in searching peaks with less than three samples
+	if (to - from < 3)
+		return;
 
-		if ((entry == max_entry) && entry->depth / 100 != last) {
-			plot_depth_sample(entry, Qt::AlignHCenter | Qt::AlignBottom, getColor(SAMPLE_DEEP));
-			last = entry->depth / 100;
+	const int half_interval = vAxis.getMinLabelDistance(hAxis);
+	const int min_depth = 2000; // in mm
+	const int min_prominence = 2000; // in mm (should this adapt to depth range?)
+	const plot_data *data = dataModel.data().entry;
+	const int max_peaks = (data[to].sec - data[from].sec) / half_interval + 1;
+	struct Peak {
+		int range_from;
+		int range_to;
+		int peak;
+	};
+	std::vector<Peak> peaks;
+	peaks.reserve(max_peaks);
+	int highest_peak = std::max_element(data + from, data + to, comp_depth) - data;
+	if (data[highest_peak].depth < min_depth)
+		return;
+	peaks.push_back(Peak{ from, to, highest_peak });
+	for (int act = 0; act < (int)peaks.size(); ++act) {
+		plot_depth_sample(data[peaks[act].peak], Qt::AlignHCenter | Qt::AlignTop, getColor(SAMPLE_DEEP));
```

Comment:
I am used to the fact that I can barely parse the C++ and have to keep re-reading things... but I'm pretty sure this one makes no sense. in line 183/184 we create a new vector `peaks`. In 188 we add one element. And now we have a `for` loop over "all the elements" in `peaks`.
What am I missing?